rpcs3/Crypto/unpkg.cpp

```diff
@@ -744,180 +747,258 @@ bool package_reader::extract_data(atomic_t<double>& sync)
-						sync += 1.;
+						if (out.write(data_span.data(), block_size) != block_size)
+						{
+							extract_success = false;
+							pkg_log.error("Failed to write file %s", path);
+							break;
+						}
+
+						if (sync.fetch_add((block_size + 0.0) / m_header.data_size) < 0.)
+						{
+							if (was_null)
+							{
+								pkg_log.error("Package installation cancelled: %s", dir);
+								out.close();
+								fs::remove_all(dir, true);
+								return;
+							}
+
+							// Cancel the installation
+							num_failures++;
+						}
 					}
-				}
 
-				if (is_buffered)
-				{
-					out = DecryptEDAT(out, name, 1, reinterpret_cast<u8*>(&m_header.klicensee), true);
-					if (!out || !fs::write_file(path, fs::rewrite, static_cast<fs::container_stream<std::vector<u8>>*>(out.release().get())->obj))
+					if (is_buffered)
 					{
-						num_failures++;
-						pkg_log.error("Failed to create file %s", path);
-						break;
+						out = DecryptEDAT(out, name, 1, reinterpret_cast<u8*>(&m_header.klicensee), true);
+						if (!out || !fs::write_file(path, fs::rewrite, static_cast<fs::container_stream<std::vector<u8>>*>(out.release().get())->obj))
+						{
+							num_failures++;
+							pkg_log.error("Failed to create file %s", path);
+							break;
+						}
 					}
-				}
 
-				if (extract_success)
-				{
-					if (did_overwrite)
+					if (extract_success)
 					{
-						pkg_log.warning("Overwritten file %s", path);
+						if (did_overwrite)
+						{
+							pkg_log.warning("Overwritten file %s", path);
+						}
+						else
+						{
+							pkg_log.notice("Created file %s", path);
+
+							if (name == "USRDIR/EBOOT.BIN" && entry.file_size > 4)
+							{
+								// Expose the creation of a bootable file
+								m_bootable_file_path = path;
+							}
+						}
 					}
 					else
 					{
-						pkg_log.notice("Created file %s", path);
-
-						if (name == "USRDIR/EBOOT.BIN" && entry.file_size > 4)
-						{
-							// Expose the creation of a bootable file
-							m_bootable_file_path = path;
-						}
+						num_failures++;
 					}
 				}
 				else
 				{
 					num_failures++;
+					pkg_log.error("Failed to create file %s", path);
 				}
-			}
-			else
-			{
-				num_failures++;
-				pkg_log.error("Failed to create file %s", path);
-			}
 
-			break;
-		}
-
-		case PKG_FILE_ENTRY_FOLDER:
-		case 0x12:
-		{
-			if (fs::create_dir(path))
-			{
-				pkg_log.notice("Created directory %s", path);
-			}
-			else if (fs::is_dir(path))
-			{
-				pkg_log.warning("Reused existing directory %s", path);
+				break;
 			}
-			else
+			default:
 			{
 				num_failures++;
-				pkg_log.error("Failed to create directory %s", path);
+				pkg_log.error("Unknown PKG entry type (0x%x) %s", entry.type, name);
+			}
 			}
-
-			break;
 		}
+	};
 
-		default:
-		{
-			num_failures++;
-			pkg_log.error("Unknown PKG entry type (0x%x) %s", entry.type, name);
-		}
-		}
-	}
+	named_thread_group workers("PKG Installer "sv, std::max<usz>(m_bufs.size(), 1) - 1, install_func);
```

Comment:
why - 1 ?
Isn't the buffer array size equivalent to the thread count?

Reply:
The calling thread alao does processing because it's a bit more efficient.